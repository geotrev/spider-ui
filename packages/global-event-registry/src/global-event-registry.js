const initContext = Symbol("#initContext")
const forEachReverse = Symbol("#forEachReverse")
const contextExists = Symbol("#contextExists")
const eventsReducer = Symbol("#eventsReducer")
const isValidContext = Symbol("#isValidContext")
const handleEvent = Symbol("#handleEvent")

/**
 * This utility is an event bus to track
 */
class GlobalEventRegistry {
  constructor() {
    this[initContext]()
    this[handleEvent] = this[handleEvent].bind(this)
  }

  // public

  register(context = {}) {
    if (!this[isValidContext](context)) return
    this[eventsReducer]("register", context.events)
    this.context.push(context)
  }

  unregister(targetId) {
    if (!this.context.length) return
    if (typeof targetId !== "string") return

    const activeContext = this.context[this.context.length - 1]

    if (activeContext.id === targetId) {
      const context = this.context.pop()
      this[eventsReducer]("unregister", context.events)
    } else {
      this[forEachReverse](this.context, (context, index) => {
        if (context.id === targetId) {
          this.context.splice(index, 1)
          this[eventsReducer]("unregister", context.events)
          return false
        }
      })
    }
  }

  get context() {
    return window.__GLOBAL_EVENT_REGISTRY__
  }

  get registry() {
    return window.__GLOBAL_EVENT_REGISTRY__.registry
  }

  // private

  [initContext]() {
    if (typeof this.context === "undefined") return

    const context = []
    const registry = {}
    context.registry = registry

    // Add it to the window. This will ensure duplicates
    // never create multiple stores.
    Object.defineProperty(window, "__GLOBAL_EVENT_REGISTRY__", {
      value: context,
    })
  }

  [forEachReverse](array, callback) {
    if (!Array.isArray(array) || !array.length) return
    for (let index = array.length - 1; index >= 0; index--) {
      const result = callback(array[index], index)
      if (result === true) break
    }
  }

  [contextExists](id) {
    let exists = false

    this[forEachReverse](this.context, (context) => {
      if (context.id !== id) return
      return true
    })

    return exists
  }

  [eventsReducer](type, events) {
    switch (type) {
      case "register": {
        events.forEach((eventType) => {
          const eventCount = this.registry[eventType]
          if (!eventCount) {
            window.addEventListener(eventType, this[handleEvent], true)
            this.registry[eventType] = 1
          } else {
            this.registry[eventType] = eventCount + 1
          }
        })
        break
      }
      case "unregister": {
        events.forEach((eventType) => {
          const eventCount = this.registry[eventType]
          if (eventCount === 1) {
            delete this.registry[eventType]
            window.removeEventListener(eventType, this[handleEvent], true)
          } else {
            this.registry[eventType] = eventCount - 1
          }
        })
        break
      }
      default: {
        break
      }
    }
  }

  [isValidContext](context) {
    return (
      Object.keys(context).length &&
      Array.isArray(context.events) &&
      context.events.length &&
      typeof context.callback === "function" &&
      typeof context.id === "string" &&
      !this[contextExists](context.id)
    )
  }

  [handleEvent](event) {
    if (!this.context.length || !this.registry) return
    const activeContext = this.context[this.context.length - 1]

    if (activeContext.events.includes(event.type)) {
      activeContext.callback(event)
    }
  }
}

const globalEventRegistry = new GlobalEventRegistry()

export { globalEventRegistry }
