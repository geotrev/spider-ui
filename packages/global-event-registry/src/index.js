export * from "./global-event-registry"
