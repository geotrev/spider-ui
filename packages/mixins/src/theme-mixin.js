import { sanitize } from "@spider-ui/utilities"

export const ThemeMixin = (Base) => {
  return class extends Base {
    constructor() {
      super()
    }

    renderTheme() {
      const theme = this.getAttribute("theme")

      if (typeof theme === "string") {
        return `<style>${sanitize(theme)}</style>`
      }

      return ""
    }
  }
}
