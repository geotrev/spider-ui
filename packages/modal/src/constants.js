export const Attributes = {}

export const ClassNames = {
  VISIBLE: "visible",
}

export const Slots = {
  TRIGGER: "[slot='trigger']",
  CONTENT: "[slot='content']",
}

export const ESCAPE_KEY = "Escape"
