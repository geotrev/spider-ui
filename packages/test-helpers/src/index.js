export * from "./timing"
export * from "./render"
