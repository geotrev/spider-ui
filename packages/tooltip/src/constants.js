export const Attributes = {
  POSITION: "position",
  MODE: "mode",
  SHOW_ARROW: "show-arrow",
  DELAY: "delay",
  DELAY_ON: "delay-on",
  DELAY_OFF: "delay-off",

  // A11y
  ARIA_DESCRIBEDBY: "aria-describedby",
  ROLE: "role",
}

export const Positions = {
  BLOCK_START: "block-start",
  BLOCK_END: "block-end",
  INLINE_START: "inline-start",
  INLINE_END: "inline-end",
}

export const Modes = {
  DARK: "dark",
  LIGHT: "light",
}

export const ClassNames = {
  ARROW: "arrow",
  VISIBLE: "visible",
  HIDDEN: "hidden",
}

export const Slots = {
  TRIGGER: "[slot='trigger']",
  CONTENT: "[slot='content']",
}

export const TIMEOUT_DELAY = 300
export const ESCAPE_KEY = "Escape"
export const TOOLTIP_TYPE = "tooltip"
