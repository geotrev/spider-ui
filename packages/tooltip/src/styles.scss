// @use "../../sass/variables" as *;
@use "@spider-ui/sass/mixins" as *;
@use "@spider-ui/sass";

.tooltip {
  position: relative;
  display: inline-block;
}

// .block-start {
//   @include mixins.create-flow-property("left", 0); // This is a fallback.
//   bottom: 100%;

//   @include mixins.create-down-arrow(
//     vars.$tooltip-arrow-offset,
//     vars.$tooltip-arrow-fill-color,
//     vars.$tooltip-arrow-border-color
//   );
// }

// .inline-end {
//   @include mixins.create-flow-property("left", 100%);
//   @include mixins.create-flow-property("margin-left", vars.$global-arrow-size);
//   margin-top: 0;
//   margin-bottom: 0;
//   bottom: auto;
//   top: 0; // This is a fallback.

//   @include mixins.create-left-arrow(
//     vars.$tooltip-arrow-offset,
//     vars.$tooltip-arrow-fill-color,
//     vars.$tooltip-arrow-border-color
//   );
// }

// .inline-start {
//   @include mixins.create-flow-property("right", 100%);
//   @include mixins.create-flow-property("left", auto);
//   @include mixins.create-flow-property("margin-right", vars.$global-arrow-size);
//   margin-top: 0;
//   margin-bottom: 0;
//   bottom: auto;
//   top: 0; // This is a fallback.

//   @include mixins.create-right-arrow(
//     vars.$tooltip-arrow-offset,
//     vars.$tooltip-arrow-fill-color,
//     vars.$tooltip-arrow-border-color
//   );
// }

// .block-end {
//   @include mixins.create-flow-property("left", 0); // This is a fallback.
//   top: 100%;
//   margin: -#{vars.$global-arrow-size} 0 0 0;
//   bottom: auto;

//   @include mixins.create-up-arrow(
//     vars.$tooltip-arrow-offset,
//     vars.$tooltip-arrow-fill-color,
//     vars.$tooltip-arrow-border-color
//   );
// }

::slotted([slot="content"]) {
  padding: 4px 6px;
  border-radius: 4px;
  border: 1px solid #cccccc;
  background: #ffffff;
  position: absolute;
  display: block;
  opacity: 0;
  transition: opacity ease-out 0.3s;
  pointer-events: none;
  word-wrap: break-word;
  white-space: nowrap;
  text-align: center;
  z-index: 1500;
  color: #333333;
  font-size: 14px;

  .is-dark & {
    border: none;
    background: #333333;
    color: #efefef;
  }

  .is-visible & {
    @include pointer--none;
    @include opacity--0;
  }
}
