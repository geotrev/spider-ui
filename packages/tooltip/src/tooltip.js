import { UpgradedElement, register } from "upgraded-element"
import { globalEventRegistry } from "@spider-ui/global-event-registry"
import { createDispatch } from "@spider-ui/dispatch"
import {
  Attributes,
  Positions,
  Modes,
  ClassNames,
  Slots,
  TIMEOUT_DELAY,
  ESCAPE_KEY,
  TOOLTIP_TYPE,
} from "./constants"
import styles from "./styles.scss"

class SpiderTooltip extends UpgradedElement {
  static get properties() {
    return {
      isVisible: {
        type: "boolean",
        default: false,
      },
    }
  }

  static get styles() {
    return styles
  }

  get classNames() {
    const getValues = (obj) => Object.keys(obj).map((key) => obj[key])
    const showArrow = this.hasAttribute(Attributes.SHOW_ARROW)
    const positionValue = this.getAttribute(Attributes.POSITION)
    const modeValue = this.getAttribute(Attributes.MODE)
    const positionValues = getValues(Positions)
    const modesValues = getValues(Modes)

    const position = positionValues.includes(positionValue)
      ? positionValue
      : Positions.BLOCK_START
    const mode = modesValues.includes(modeValue) ? modeValue : Modes.DARK
    const arrow = showArrow ? ClassNames.ARROW : ""
    const visible = this.isVisible ? ClassNames.VISIBLE : ClassNames.HIDDEN

    return { position, mode, arrow, visible }
  }

  constructor() {
    super()
    this.handleOpen = this.handleOpen.bind(this)
    this.handleClose = this.handleClose.bind(this)
    this.handleKeydown = this.handleKeydown.bind(this)
    this.removeTimeout = this.removeTimeout.bind(this)
  }

  elementDidMount() {
    // set nodes
    const trigger = this.querySelector(Slots.TRIGGER)
    const content = this.querySelector(Slots.CONTENT)

    // set attributes
    trigger.setAttribute(Attributes.ARIA_DESCRIBEDBY, this.elementId)
    content.id = this.elementId
    content.setAttribute(Attributes.ROLE, "tooltip")

    this.addOpenListeners()
    this.addCloseListeners()
  }

  elementWillUnmount() {
    this.handleOpen = null
    this.handleClose = null
    this.handleKeydown = null
    this.removeTimeout = null

    globalEventRegistry.unregister(this.elementId)
  }

  elementDidUpdate() {
    if (this.isVisible) {
      if (this.isInlineTooltip()) {
        this.alignTooltip("height")
      } else {
        this.alignTooltip("width")
      }

      globalEventRegistry.register({
        types: ["keydown"],
        id: this.elementId,
        handler: this.handleKeydown,
      })
    } else {
      globalEventRegistry.unregister(this.elementId)
    }
  }

  setDelay() {
    const delayOn = this.getAttribute(Attributes.DELAY_ON)
    const delayOff = this.getAttribute(Attributes.DELAY_OFF)
    const delay = this.getAttribute(Attributes.DELAY)

    this.delayOn = parseInt(delayOn || delay) || TIMEOUT_DELAY
    this.delayOff = parseInt(delayOff || delay) || TIMEOUT_DELAY
  }

  addOpenListeners() {
    this.addEventListener("focusin", this.handleOpen)
    this.addEventListener("focusin", this.handleOpen)
    this.addEventListener("mouseover", this.handleOpen)
    this.addEventListener("mouseover", this.handleOpen)
  }

  addCloseListeners() {
    this.addEventListener("focusout", this.handleClose)
    this.addEventListener("focusout", this.handleClose)
    this.addEventListener("mouseout", this.handleClose)
    this.addEventListener("mouseout", this.handleClose)
  }

  addOpenCancelListeners() {
    this.addEventListener("mouseout", this.removeTimeout)
    this.addEventListener("focusout", this.removeTimeout)
    this.addEventListener("mouseout", this.removeTimeout)
  }

  addCloseCancelListeners() {
    this.addEventListener("mouseover", this.removeTimeout)
    this.addEventListener("focusin", this.removeTimeout)
    this.addEventListener("mouseover", this.removeTimeout)
  }

  removeOpenCancelListeners() {
    this.removeEventListener("mouseout", this.removeTimeout)
    this.removeEventListener("focusout", this.removeTimeout)
    this.removeEventListener("mouseout", this.removeTimeout)
  }

  removeCloseCancelListeners() {
    this.removeEventListener("mouseover", this.removeTimeout)
    this.removeEventListener("focusin", this.removeTimeout)
    this.removeEventListener("mouseover", this.removeTimeout)
  }

  handleKeydown(event) {
    if (this.isVisible && event.key === ESCAPE_KEY) {
      this.removeTimeout()
      this.removeCloseCancelListeners()
      this.isVisible = false
      this.dispatchEvent(createDispatch(TOOLTIP_TYPE, { visible: false }))
    }
  }

  removeTimeout() {
    if (!this.isConnected) return
    if (!this.timeout) return
    clearTimeout(this.timeout)
    this.timeout = null
  }

  handleOpen() {
    if (this.isVisible) return
    const _this = this

    this.timeout = setTimeout(() => {
      if (!_this.isConnected) return
      _this.removeOpenCancelListeners()
      _this.timeout = null
      _this.isVisible = true
      _this.dispatchEvent(createDispatch(TOOLTIP_TYPE, { visible: true }))
    }, this.delayOn)

    this.addOpenCancelListeners()
  }

  handleClose() {
    if (!this.isVisible) return
    const _this = this

    this.timeout = setTimeout(() => {
      if (!_this.isConnected) return
      _this.removeCloseCancelListeners()
      _this.timeout = null
      _this.isVisible = false
      _this.dispatchEvent(createDispatch(TOOLTIP_TYPE, { visible: false }))
    }, this.delayOff)

    this.addCloseCancelListeners()
  }

  alignTooltip(dimension) {
    const trigger = this.querySelector(Slots.TRIGGER)
    const content = this.querySelector(Slots.CONTENT)
    const triggerSize = this.getSize(trigger, dimension)
    const tooltipSize = this.getSize(content, dimension)
    const triggerIsBigger = triggerSize > tooltipSize

    const offset = triggerIsBigger
      ? (triggerSize - tooltipSize) / 2
      : (tooltipSize - triggerSize) / -2

    if (dimension === "height") {
      content.style.top = `${offset}px`
    } else {
      content.style.left = `${offset}px`
    }
  }

  getSize(element, property) {
    return Math.floor(element.getBoundingClientRect()[property])
  }

  isInlineTooltip() {
    const position = this.getAttribute(Attributes.POSITION)
    return [Positions.INLINE_START, Positions.INLINE_END].includes(position)
  }

  render() {
    const { position, mode, arrow, visible } = this.classNames
    this.setDelay()

    return `
      <div class="tooltip ${visible} ${position} ${mode} ${arrow}">
        <slot name="trigger"></slot>
        <slot name="content"></slot>
      </div>
    `
  }
}

register("spider-tooltip", SpiderTooltip)
