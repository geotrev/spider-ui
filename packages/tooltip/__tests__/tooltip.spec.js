import "../src"

describe("@spider-ui/tooltip", () => {
  it("needs tests", () => {})
})
