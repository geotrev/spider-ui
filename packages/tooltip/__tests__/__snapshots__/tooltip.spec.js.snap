// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<spider-tooltip> default configuration renders default state 1`] = `
<div
  class="tooltip hidden block-start dark "
>
  <slot
    name="trigger"
  />
  <slot
    name="content"
  />
</div>
`;
